docs/celestia-architecture/adr-011-sequence-aware-cat.md

```diff
@@ -0,0 +1,104 @@
+# ADR 011: Sequence and priority ordering CAT mempool
+
+## Changelog
+
+- 2025-09-04: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+The CAT mempool previously ordered transactions strictly based on priority (fee / gas). This is incongruent with the replay protection mechanism which depends on the order of transactions by a signer having a monotonically increasing sequence number. This becomes more apparent with multiple transactions by the same signer in the same block.
+
+We need a mempool ordering algorithm that respects per-signer sequencing constraints while still maximising revenue through choosing the highest paid tranasctions (per byte or gas used).
+
+## Decision
+
+Group transactions by signer into signer-bundled sets ("tx sets") and order based on average weighted priority per set. Within tx set order transactions in ascending sequence order. Block building and recheckTx iterates through transactions in those order. This will preserve sequence ordering per signer while maximising tx fees per block.
```

Comment:
[question] Is it possible for an attacker to submit two txs:
1. tx 1 with sequence 1 and a low priority
2. tx 2 with sequence 2 and a high priority

such that the second tx increases the average weighted priority of the tx set such that tx 1 gets included on-chain, however once tx 1 gets included on-chain it invalidates tx 2 so tx 2 can no longer get on chain. The end result is that the signer got tx 1 included on-chain ahead of competing txs that would have paid more in fees?

Update: oops disregard, I see Rachid raised the same concern.

Reply:
Responded to it in rachid's comment